File: src/leetcode/python/BinaryTreePaths_257.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreePaths_257(object):
    # @param {TreeNode} root
    # @return {string[]}
    def binaryTreePaths(self, root):
        res, stack = [], []
        if not root:
            return res
        stack.append([root, str(root.val)])
        while stack:
            node, val = stack.pop()
            if not node.left and not node.right:
                res.append(val)
            if node.right:
                stack.append([node.right, val+'->'+str(node.right.val)])
            if node.left:
                stack.append([node.left, val+'->'+str(node.left.val)])
        return res

File: src/leetcode/java/BinaryTreePaths_257.java
package leetcode.java;

import java.util.ArrayList;
import java.util.List;

public class BinaryTreePaths_257 {
  private List<String> result = new ArrayList<>();
  private List<Integer> list = new ArrayList<>();

  public List<String> binaryTreePaths(TreeNode root) {
    if (root == null) return result;
    dfs(root);
    return result;
  }

  private void dfs(TreeNode node) {
    if (node == null) return;
    list.add(node.val);
    if (node.left == null && node.right == null) {
      result.add(convertToString(list));
      list.remove(list.size() - 1);
      return;
    }
    dfs(node.left);
    dfs(node.right);
    list.remove(list.size() - 1);
  }

  private String convertToString(List<Integer> root2Leaf) {
    StringBuilder builder = new StringBuilder();
    for (int i = 0; i < root2Leaf.size(); i++) {
      if (i == root2Leaf.size() - 1) builder.append(root2Leaf.get(i));
      else builder.append(root2Leaf.get(i)).append("->");
    }
    return builder.toString();
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
